New Python application for generating Ramp card activity and distributing via email

lewismc opened the pull request:
Hi @craigmcc this is a work in progress for the monthly reporting automation we discussed the last couple of weeks.

The new Python application is proposed to be located in `apps/statement-distributor` but I suspect we will wish to rename this directory. Please let me know your thoughts.
I propose to update the app once the email information is available from the Ramp DB. At that stage we can go through it in more detail.

This PR also proposes a change to `apps/ramp-refresh/src/Refreshers.ts` which I believe address a foreign key constraint issue when refreshing user data from the Ramp API. In the User database schema, there's a self-referential foreign key relationship where manager_id points to another user's ID. When users are being inserted/updated, if a user's manager hasn't been created yet in the database, it would cause a foreign key constraint violation.

This PR proposes a two-pass approach which 

1.  Inserts all users without manager relationships, then
2. updates manager relationships

This should eliminate foreign key errors because all users exist before relationships are created. It can be run multiple times safely. It shows progress for both passes. It will works even if managers are listed after their reports in the API response.

If you want me to remove this tangential change into a separate PR please let me know.

lewismc committed:
New Python application for generating Ramp card activity and distributing via email.

craigmcc commented:
I am heading to an appointment but will look at these this afternoon.  Just
two quick points for now:
* Turborepo and pnpm assume anything in “apps” or “packages” is typescript
stuff, so Python code will need to go in some other top level directory.
* The user/manager stuff sounds like a good idea, but let’s keep PRs single
focused.

Craig

On Fri, Nov 21, 2025 at 10:58 AM Lewis John McGibbney <
***@***.***> wrote:

> Hi @craigmcc <https://github.com/craigmcc> this is a work in progress for
> the monthly reporting automation we discussed the last couple of weeks.
>
> The new Python application is proposed to be located in
> apps/statement-distributor but I suspect we will wish to rename this
> directory. Please let me know your thoughts.
> I propose to update the app once the email information is available from
> the Ramp DB. At that stage we can go through it in more detail.
>
> This PR also proposes a change to apps/ramp-refresh/src/Refreshers.ts
> which I believe address a foreign key constraint issue when refreshing user
> data from the Ramp API. In the User database schema, there's a
> self-referential foreign key relationship where manager_id points to
> another user's ID. When users are being inserted/updated, if a user's
> manager hasn't been created yet in the database, it would cause a foreign
> key constraint violation.
>
> This PR proposes a two-pass approach which
>
>    1. Inserts all users without manager relationships, then
>    2. updates manager relationships
>
> This should eliminate foreign key errors because all users exist before
> relationships are created. It can be run multiple times safely. It shows
> progress for both passes. It will works even if managers are listed after
> their reports in the API response.
>
> If you want me to remove this tangential change into a separate PR please
> let me know.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/craigmcc/apache-turborepo/pull/4
> Commit Summary
>
>    - 5f3b290
>    <https://github.com/craigmcc/apache-turborepo/pull/4/commits/5f3b29091369698f809f604d72e523a4878e3ff8>
>    New Python application for generating Ramp card activity and distributing
>    via email.
>
> File Changes
>
> (6 files <https://github.com/craigmcc/apache-turborepo/pull/4/files>)
>
>    - *M* .gitignore
>    <https://github.com/craigmcc/apache-turborepo/pull/4/files#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947>
>    (27)
>    - *M* apps/ramp-refresh/src/Refreshers.ts
>    <https://github.com/craigmcc/apache-turborepo/pull/4/files#diff-75cc4c4c0c3f209bb8d54950598304c15a615b536c5cde1a8deceb292a7439d6>
>    (45)
>    - *A* apps/statement-distributor/README.md
>    <https://github.com/craigmcc/apache-turborepo/pull/4/files#diff-81ce1541e149608b0a052e1d980f6992bd4371af7210aff73277eb56b65475ce>
>    (240)
>    - *A* apps/statement-distributor/config.example.json
>    <https://github.com/craigmcc/apache-turborepo/pull/4/files#diff-7c98760d8996473427df31d042b2ecf3ac54afaef89a2218331df1856317dd36>
>    (37)
>    - *A* apps/statement-distributor/requirements.txt
>    <https://github.com/craigmcc/apache-turborepo/pull/4/files#diff-15e9e92390cf5281af1a8315c7fc639acb5e0dad3e3f289292c2036ebba72f11>
>    (2)
>    - *A* apps/statement-distributor/statement_distributor.py
>    <https://github.com/craigmcc/apache-turborepo/pull/4/files#diff-db053c11c5467f6ab6ee4258f5a7ff99b978318817398e2afc6647426b6789b5>
>    (434)
>
> Patch Links:
>
>    - https://github.com/craigmcc/apache-turborepo/pull/4.patch
>    - https://github.com/craigmcc/apache-turborepo/pull/4.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/craigmcc/apache-turborepo/pull/4>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA7ZLXALAAULB5RV4JA7Z3355OFRAVCNFSM6AAAAACM22ZOZKVHI2DSMVQWIX3LMV43ASLTON2WKOZTGY2TENZWGU2TOOI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

craigmcc reviewed (changes requested):
As mentioned in email, the changes related to Refreshers.ts should be in a separate PR.  They look good, and I don't think there are any other cases where we have potential issues like this, because of the ordering of table refreshes.

On the Python app:
* It definitely is not complete yet, but looks like a good start.  However, it *cannot* be in the apps directory, because Turborepo (and pnpm) assume all such content, as well as anything in packages, belongs to them.  See the "packages" tag in pnpm-workspace.yaml to see what directories they pay attention to.
* We'll need to do bill.com stuff as well.  I suspect we'll want a separate app for that, because they are totally independent things.  And they should probably be named that way - ramp_statement_distributor.py and bill_statement_distributor.py for example.
* Does the Python universe have a concept like .env files to set environment variables?  If so, we should definitely go that way, because it's needed for all the Typescript apps anyway.
* Note that treasury@apache.org is not a valid email address unless we ask infra to make one.  However, treasurer@apache.org is valid, and is a group that includes both of us (as well as potential other folks) already.
* Does it make sense to put the mapping of departments to emails in the AccountGroups.json file, only for those that need such forwarding?  That way, it could be shared for both bill and ramp reporting.
* It looks like we will be relying on port 587 to send SMTP messages for us.  I have such a thing on my Mac, and I assume Windows does too, but we'll want to verify that actually works with our Apache email infrastructure receiving them.

I like the idea of a README.md file for each app, and will look at doing that for the various Typescript apps as well, pulling the details out of the global README file and just pointing at them where relevant.

lewismc committed:
Merge branch 'main' into statement-distributor

lewismc committed:
New Python application for generating Ramp card activity and distributing via email

lewismc committed:
New Python application for generating Ramp card activity and distributing via email

lewismc committed:
New Python application for generating Ramp card activity and distributing via email

lewismc commented:
@craigmcc OK I finally made an update to this PR. Here's a rundown
* renamed `statement-distributor` to `ramp-statement-distributor`
* relocated `ramp-statement-distributor` to a new top-level directory named `distributors` for non-Turporepo services.
* Linked to the ramp-statement-distributor README from the top-level README
* introduced [pyenv](https://github.com/pyenv/pyenv) for python version management and pinned the python version at 3.11.0.
* Introduced a setup.sh script which will use the pyenv and virtual environment to install all dependencies so it does not affect the host python environment. 
* Uses correct treasurer@ email address

Are you OK with the department emails being statically defined in `distributors/ramp-statement-distributor/config.example.json`? Or should these be pulled dynamically from `ramp-lookup`?
Thanks

lewismc commented:
# 2025-12-18 FEEDBACK

* ~~Add month range functionality `--month` parameter~~
* Fetch account group emails from `AccountGroups.json` rather than `config.json` (retain --account-groups flag and functionality).
* Add functionality to subset account groups by account group name (one or more comma separated) e.g. `python ramp_statement_distributor.py --config config.json --account-groups /path/to/AccountGroups.json --account-groups-name TAC,Treasury`

lewismc committed:
New Python application for generating Ramp card activity and distributing via email

lewismc commented:
Hi @craigmcc with the most recent commit, you can now fetch transactions by month range. Note the following examples

### Basic Usage (Previous Month)

By default, the tool processes the previous month's data:

```bash
python ramp_statement_distributor.py --config config.json
```

### Specify a Specific Month

To process a specific month:

```bash
python ramp_statement_distributor.py --config config.json --month 2024-11
```

### Specify a Month Range

To process multiple months at once:

```bash
python ramp_statement_distributor.py --config config.json --from-month 2024-01 --to-month 2024-12
```

lewismc committed:
Merge branch 'main' into statement-distributor

lewismc committed:
New Python application for generating Ramp card activity and distributing via email

lewismc commented:
Hi @craigmcc OK so after a fair amount of work this PR is now at a demo'able stage. I've run it end-to-end and am fairly happy with it.
Major improvements/adjustments from the last time I demo'd it.
* transactions are read from the local Ramp SQLite DB, which removes the dependency on ramp lookup service. 
* automatic CSV merging for multi-month ranges
* daily rotating logs, configurable via `config.json`
* a **_Summary Report_** can be emailed to Treasury whenever the tool is run, this is configurable as well

Look forward to demo'ing this tomorrow. Thanks

lewismc committed:
New Python application for generating Ramp card activity and distributing via email

lewismc committed:
New Python application for generating Ramp card activity and distributing via email

lewismc commented:
**LOCAL TESTING**


<img width="733" height="365" alt="Screenshot 2026-01-07 at 22 52 58" src="https://github.com/user-attachments/assets/532cf689-5b89-46bf-94f3-011b228e9adf" />
<img width="713" height="315" alt="Screenshot 2026-01-07 at 22 53 30" src="https://github.com/user-attachments/assets/a09b6110-73cc-4e2a-9fde-1d6ed2906a0d" />

lewismc committed:
New Python application for generating Ramp card activity and distributing via email

lewismc commented:
@craigmcc I implemented all the changes we discussed yesterday. New CLI 
```
python ramp_statement_distributor.py                                                
usage: ramp_statement_distributor.py [-h] [--config CONFIG] [--departments DEPARTMENTS] [--list-departments] [--from-date FROM_DATE]
                                     [--to-date TO_DATE] [--send-emails]

Generate and distribute credit card statements for specified date ranges

options:
  -h, --help            show this help message and exit
  --config CONFIG       Path to configuration JSON file (default: config.json)
  --departments DEPARTMENTS
                        Comma-separated list of departments to process (case-insensitive). If not specified, all departments are processed.
  --list-departments    List all available departments and exit
  --from-date FROM_DATE
                        Start date in YYYY-MM-DD format (default: first day of previous month)
  --to-date TO_DATE     End date in YYYY-MM-DD format (default: last day of previous month)
  --send-emails         Actually send emails (default: dry-run mode, emails are not sent)
```

My default invoking the tool without any command print help. Sending emails is deactivated by default (to prevent unintended email distribution) and can be turned on by passing `--send-emails`. The new `--from-date` and `--to-date` flags facilitate more precise range queries.
I added the treasury email address to the email template. I also made some improvements to the summary email format which will include reasons for errors if any occur.

lewismc committed:
New Python application for generating Ramp card activity and distributing via email
